### DISCUSSION TITLE
API Docs for DI.FM?

### DISCUSSION THREAD

**Discussion by jonathanknez** [acaloiaro/di-tui#25] [Q&A] [answered]
I am interested in learning more about the DI.FM web API. Your difm.go lays it out nicely. Did you reverse engineer it or is there a published API doc? My google search didn't even turn up the audioaddict.com API which looks very useful. Just wondering if there is a public source for DI.FM API information before I reach out to the company itself. Thanks.

**Comment by acaloiaro** [comment, marked as answer]
I don’t think you’ll have much luck finding any docs or getting in touch with them about docs, but don’t take my word for it. 

The API isn’t terribly complicated, but they’ve recently placed cloudflare in front of their web application, which unfortunately blocks any vanilla web requests to their endpoints since it checks whether clients have JavaScript available. For this reason, logging in via username and password with di-tui no longer works, and I have no intention of trying to skirt it. 

With that said, their APIs can be reverse engineered pretty trivially with a browser’s network tab when logged in. It’s not part of this public project, but I was able to reverse engineer their “on demand” audio API that way, but since those APIs don’t support MP3 audio, and don’t use any codecs with Go decoders, I opted to keep with the streaming API. 

This code base, small as it is, is probably one of the more complete client references I was able to find.

**Reply by jonathanknez** [reply]
Thank you for your response. Reverse engineering it is then, which is made substantially easier by referring to your excellent work. Take care.

**Reply by acaloiaro** [reply]
Thanks, and good luck!
